test/pkg/environment/clustermonitor.go

```diff
@@ -0,0 +1,164 @@
+	"knative.dev/pkg/logging"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+// ClusterMonitor is used to monitor the cluster state during a running test
+type ClusterMonitor struct {
+	ctx        context.Context
+	kubeClient client.Client
+
+	mu                 sync.RWMutex
+	recordings         []recording
+	nodesSeen          sets.String
+	numberNodesAtReset int
+}
+type recording struct {
+	nodes v1.NodeList
+	pods  v1.PodList
+}
+
+func NewClusterMonitor(ctx context.Context, kubeClient client.Client) *ClusterMonitor {
+	m := &ClusterMonitor{
+		ctx:        ctx,
+		kubeClient: kubeClient,
+		nodesSeen:  sets.NewString(),
+	}
+	m.Reset()
+
+	go func() {
+		for {
+			select {
+			case <-ctx.Done():
+				return
+			case <-time.After(1 * time.Second):
+				m.poll()
+			}
+		}
+	}()
+	return m
+}
+
+// Reset resets the cluster monitor prior to running a test.
+func (m *ClusterMonitor) Reset() {
+	m.mu.Lock()
+	m.recordings = nil
+	m.nodesSeen = map[string]sets.Empty{}
+	m.mu.Unlock()
+	m.poll()
+	m.numberNodesAtReset = len(m.nodesSeen)
+}
+
+// RestartCount returns the containers and number of restarts for that container for all containers in the pods in the
+// given namespace
+func (m *ClusterMonitor) RestartCount(namespace string) map[string]int {
+	m.poll()
+
+	m.mu.RLock()
+	defer m.mu.RUnlock()
+	restarts := map[string]int{}
+	last := m.recordings[len(m.recordings)-1]
+	for _, pod := range last.pods.Items {
+		if pod.Namespace != namespace {
+			continue
+		}
+		for _, cs := range pod.Status.ContainerStatuses {
+			name := fmt.Sprintf("%s/%s", pod.Name, cs.Name)
+			restarts[name] = int(cs.RestartCount)
+		}
+	}
+	return restarts
+}
+
+// NodeCount returns the current number of nodes
+func (m *ClusterMonitor) NodeCount() int {
+	m.poll()
+	m.mu.RLock()
+	defer m.mu.RUnlock()
+	last := m.recordings[len(m.recordings)-1]
+	return len(last.nodes.Items)
+}
+
+// NodeCountAtReset returns the number of nodes that were running when the monitor was last reset, typically at the
+// beginning of a test
+func (m *ClusterMonitor) NodeCountAtReset() interface{} {
+	m.mu.RLock()
+	defer m.mu.RUnlock()
+	return m.numberNodesAtReset
+}
+
+// TotalNodesSeen returns the total number of unique nodes ever seen since the last reset.
+func (m *ClusterMonitor) TotalNodesSeen() int {
+	m.poll()
+	m.mu.RLock()
+	defer m.mu.RUnlock()
+	return len(m.nodesSeen)
+}
+
+// CreatedNodes returns the number of nodes created since the last reset
+func (m *ClusterMonitor) CreatedNodes() int {
+	return m.TotalNodesSeen() - m.numberNodesAtReset
+}
+
+// RunningPods returns the number of running healthy pods matching the given selector
+func (m *ClusterMonitor) RunningPods(selector labels.Selector) int {
+	m.poll()
+	m.mu.RLock()
+	defer m.mu.RUnlock()
+	last := m.recordings[len(m.recordings)-1]
+	count := 0
+	for _, pod := range last.pods.Items {
+		if pod.Status.Phase != v1.PodRunning {
+			continue
+		}
+		isReady := false
+		for _, cond := range pod.Status.Conditions {
```

Comment:
You can skip the phase check if you're checking conditions. I'd do one or the other.